### PR TITLE
feature: Added month and day view

### PR DESCRIPTION
**Pull request by ParthBaraiya**
- Added month view
- Added day view
- Added example to demonstrate month view and day view.

### CODE REVIEW

File: lib/src/calendar_event_data.dart [hunk 1]
```diff
@@ -8,28 +8,54 @@ class CalendarEventData<T> {
   /// Specifies date on which all these events are.
   final DateTime date;
 
-  /// List of events on [CalendarEventData.date].
+  /// Defines the start time of the event.
+  /// [endTime] and [startTime] will defines time on same day.
```
Comment (BirjuVachhani) [inline lib/src/calendar_event_data.dart:12]: replace `will defines` with `will define` or `defines`

File: lib/src/calendar_event_data.dart [hunk 2]
```diff
@@ -8,28 +8,54 @@ class CalendarEventData<T> {
...
+
+  /// Description of the event.
+  final String description;
+
+  /// List of events on [date].
   final T event;
 
   /// Stores all the events on [date]
   CalendarEventData({
     @required this.date,
```
Comment (BirjuVachhani) [inline lib/src/calendar_event_data.dart:32]: Why we are not using null-safety?

File: lib/src/components/common_components.dart [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  /// Date of month/day.
+  final DateTime date;
+
+  /// Provides string to display as title.
+  final StringProvider dateStringBuilder;
+
+  /// Common header for month and day view In this header user can define format
+  /// in which date will be displayed by providing [dateStringBuilder] function.
+  ///
+  const CalendarPageHeader({
```
Comment (BirjuVachhani) [inline lib/src/components/common_components.dart:25]: Don't forget to mark things as `required`

File: lib/src/components/month_view_components.dart [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+      color: backgroundColor,
+      child: Column(
+        mainAxisAlignment: MainAxisAlignment.start,
+        crossAxisAlignment: CrossAxisAlignment.center,
+        mainAxisSize: MainAxisSize.max,
+        children: [
+          SizedBox(
+            height: 5.0,
+          ),
+          shouldHighlight
```
Comment (BirjuVachhani) [inline lib/src/components/month_view_components.dart:97]: `Visibility` can be used here.

File: lib/src/components/month_view_components.dart [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+                )
+              : Text(
+                  "${date.day}",
+                ),
+          if (events.isNotEmpty)
+            Expanded(
+                child: Container(
+              margin: EdgeInsets.only(top: 5.0),
+              clipBehavior: Clip.antiAlias,
+              decoration: BoxDecoration(),
```
Comment (BirjuVachhani) [inline lib/src/components/month_view_components.dart:112]: Is there any reason for empty `BoxDecoration()`?

File: lib/src/components/month_view_components.dart [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+              child: SingleChildScrollView(
+                physics: BouncingScrollPhysics(),
+                child: Column(
+                  mainAxisSize: MainAxisSize.min,
+                  crossAxisAlignment: CrossAxisAlignment.center,
+                  mainAxisAlignment: MainAxisAlignment.start,
+                  children: List.generate(
+                      events.length,
+                      (index) => GestureDetector(
+                            onTap: () {
```
Comment (BirjuVachhani) [inline lib/src/components/month_view_components.dart:122]: Use expression body

File: lib/src/components/month_view_components.dart [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+                                      style: TextStyle(
+                                        color: Colors.white.withAlpha(240),
+                                        fontSize: 12,
+                                      ),
+                                    ),
+                                  ),
+                                ],
+                              ),
+                            ),
+                          )),
```
Comment (BirjuVachhani) [inline lib/src/components/month_view_components.dart:152]: Format it properly.

File: lib/src/constants.dart [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+import 'dart:ui';
+
+class Constants {
+  Constants._();
+
+  static final Random _random = Random();
+  static final int _maxColor = 256;
+
+  /// minimum and maximum dates are 100,000,000 days before and after epochDate
+  ///
```
Comment (BirjuVachhani) [inline lib/src/constants.dart:11]: Redundant line

File: lib/src/constants.dart [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+  static final DateTime epochDate = DateTime(1970, 1, 1);
+  static final DateTime maxDate = DateTime(275760, 1, 1);
```
Comment (BirjuVachhani) [inline lib/src/constants.dart:13]: Make them `const`

File: lib/src/day_view/_internal_day_view_page.dart [hunk 10]
```diff
@@ -0,0 +1,218 @@
+import 'dart:async';
+
+import 'package:flutter/material.dart';
+import 'package:flutter_calendar_page/flutter_calendar_page.dart';
+import 'package:flutter_calendar_page/src/extensions.dart';
+
+import '../painters.dart';
+import 'modals.dart';
+
+class InternalDayViewPage<T> extends StatelessWidget {
```
Comment (BirjuVachhani) [inline lib/src/day_view/_internal_day_view_page.dart:10]: This class requires documentation

File: lib/src/day_view/_internal_day_view_page.dart [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+        child: eventTileBuilder(
+          date,
+          events[index].events,
+          Rect.fromLTWH(
+              events[index].left,
+              events[index].top,
+              width - events[index].right - events[index].left,
+              height - events[index].bottom - events[index].top),
+          events[index].startDuration,
+          events[index].endDuration,
```
Comment (BirjuVachhani) [inline lib/src/day_view/_internal_day_view_page.dart:151]: Make this named arguments.

File: lib/src/day_view/_internal_day_view_page.dart [hunk 12]
```diff
@@ -0,0 +1,218 @@
...
+        ),
+      );
+    });
+  }
+}
+
+class _LiveTimeIndicator extends StatefulWidget {
```
Comment (BirjuVachhani) [inline lib/src/day_view/_internal_day_view_page.dart:158]: Requires documentation

File: lib/src/day_view/_internal_day_view_page.dart [hunk 13]
```diff
@@ -0,0 +1,218 @@
...
+  void dispose() {
+    _timer?.cancel();
+    super.dispose();
+  }
+
+  void setTimer() {
+    if (mounted)
+      setState(() {
+        _currentDate = DateTime.now();
+        _timer = Timer(Duration(minutes: 1), setTimer);
```
Comment (BirjuVachhani) [inline lib/src/day_view/_internal_day_view_page.dart:200]: This won't be accurate if you start at half a min or so. Maybe we could improve it. Find out other ways of doing this.

File: lib/src/day_view/day_view_page.dart [hunk 14]
```diff
@@ -0,0 +1,69 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_calendar_page/flutter_calendar_page.dart';
+import 'package:flutter_calendar_page/src/day_view/_internal_day_view_page.dart';
+import 'package:flutter_calendar_page/src/extensions.dart';
+
+import 'modals.dart';
+
+export 'modals.dart';
+
+class DayViewPage<T> extends StatelessWidget {
```
Comment (BirjuVachhani) [inline lib/src/day_view/day_view_page.dart:10]: Requires documentation

File: lib/src/day_view/modals.dart [hunk 15]
```diff
@@ -0,0 +1,18 @@
+import 'package:flutter/material.dart';
+
+class HourIndicatorSettings {
```
Comment (BirjuVachhani) [inline lib/src/day_view/modals.dart:3]: Requires documentation

File: lib/src/event_arrangers/event_arrangers.dart [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+  ///
+  const EventArranger();
+
+  /// This method will arrange all the events in and return List of [OrganizedCalendarEventData].
+  ///
+  List<OrganizedCalendarEventData<T>> arrange(List<CalendarEventData<T>> events,
+      double height, double width, double heightPerMinute);
+}
+
+class OrganizedCalendarEventData<T> {
```
Comment (BirjuVachhani) [inline lib/src/event_arrangers/event_arrangers.dart:26]: Requires documentation

File: lib/src/extensions.dart [hunk 17]
```diff
@@ -6,4 +8,69 @@ extension DateTimeExtensions on DateTime {
...
+
+  List<DateTime> get datesOfMonths {
+    List<DateTime> monthDays = [];
+    for (int i = 1, start = 1; i < 7; i++, start += 7) {
+      monthDays.addAll(DateTime(this.year, this.month, start).datesOfWeek);
+    }
+    return monthDays;
+  }
+
+  String get formatted {
```
Comment (BirjuVachhani) [inline lib/src/extensions.dart:62]: use expression body

File: lib/src/extensions.dart [hunk 18]
```diff
@@ -6,4 +8,69 @@ extension DateTimeExtensions on DateTime {
...
+    return "${this.month}-${this.year}";
+  }
+
+  int get getTotalMinutes => this.hour * 60 + this.minute;
+
+  DateTime fromMinutes([int totalMinutes = 0]) => DateTime(
```
Comment (BirjuVachhani) [inline lib/src/extensions.dart:68]: Shouldn't this be named `copyWithMinutes` or something like that?

File: lib/src/painters.dart [hunk 19]
```diff
@@ -0,0 +1,84 @@
+import 'package:flutter/material.dart';
+
+class HourLinePainter extends CustomPainter {
```
Comment (BirjuVachhani) [inline lib/src/painters.dart:3]: Requires documentation

File: lib/src/painters.dart [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+    HourLinePainter oldPainter = oldDelegate;
+    return oldPainter.lineColor != this.lineColor ||
+        oldPainter.offset != this.offset ||
+        this.lineHeight != oldPainter.lineHeight ||
+        this.minuteHeight != oldPainter.minuteHeight ||
+        this.showVerticalLine != oldPainter.showVerticalLine;
+  }
+}
+
+class CurrentTimeLinePainter extends CustomPainter {
```
Comment (BirjuVachhani) [inline lib/src/painters.dart:47]: Requires documentation

**Review by BirjuVachhani — CHANGES_REQUESTED** [review]

File: lib/src/components/month_view_components.dart (hunk 5 above)
Comment (ParthBaraiya) [reply lib/src/components/month_view_components.dart:112]: Getting this error if I do not provide box decoration.

**The following assertion was thrown building FilledCell(dirty):
'package:flutter/src/widgets/container.dart': Failed assertion: line 273 pos 15: 'decoration != null || clipBehavior == Clip.none': is not true.**

File: lib/src/constants.dart (hunk 9 above)
Comment (ParthBaraiya) [reply lib/src/constants.dart:13]: Can not assign DateTime object as constant.

Tried using,

`static const DateTime epochDate = const DateTime(1970, 1,1);
 static const DateTime maxDate = DateTime(275760, 1,1)`

none of them are working.

File: lib/src/components/common_components.dart (hunk 3 above)
Comment (ParthBaraiya) [reply lib/src/components/common_components.dart:25]: All the required parameters are marked as required. other callbacks are not required.

File: lib/src/components/month_view_components.dart (hunk 4 above)
Comment (ParthBaraiya) [reply lib/src/components/month_view_components.dart]: Instead of replacing a widget now we are changing background color only.

**Review by BirjuVachhani — APPROVED** [review]
LGTM

**Pull request merged**